src/main/java/nextstep/subway/domain/Sections.java

```diff
@@ -1,29 +1,162 @@
+                sections.add(new Section(line, upStation, downStation, distance));
+                sections.add(new Section(line, section.getUpStation(), upStation, section.getDistance() - distance));
+                sections.remove(section);
+                return;
+            }
+        }
+
+        for (Section section : sections) {
+            if (section.isDownStation(upStation) || section.isUpStation(downStation)) {
+                sections.add(new Section(line, upStation, downStation, distance));
+                return;
+            }
+        }
+
         sections.add(new Section(line, upStation, downStation, distance));
     }
 
+    private void checkSameSectionExist(Station upStation, Station downStation) {
+        for (Section section : sections) {
+            section.hasSameStations(upStation, downStation);
+        }
+    }
+
+    public void checkStationsExist(Station upStation, Station downStation) {
+        if (sections.isEmpty()) {
+            return;
+        }
+
+        for (Section section : sections) {
+            if (section.contains(upStation, downStation)) {
+                return;
+            }
+        }
+
+        throw new SectionException();
+    }
+
+    private void checkAvailableDistance(int distance, Section section) {
+        if (section.getDistance() <= distance) {
+            throw new SectionException();
+        }
+    }
+
+    public List<Station> getStations() {
+        List<Station> allStations = allStations();
+        if (allStations.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        Station upStation = findFirstStation(allStations);
+        List<Station> stations = new ArrayList<>(List.of(upStation));
+
+        while (isNotLastStation(upStation)) {
+            upStation = findNextStation(upStation, stations);
+        }
+
+        return stations;
+    }
+
+    private Station findNextStation(Station upStation, List<Station> stations) {
+        for (Section section : sections) {
+            if (section.isUpStation(upStation)) {
+                stations.add(section.getDownStation());
+                upStation = section.getDownStation();
+                break;
+            }
+        }
+        return upStation;
+    }
+
+    private boolean isNotLastStation(Station firstStation) {
+        for (Section section : sections) {
+            if (section.isUpStation(firstStation)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private List<Station> allStations() {
+        if (sections.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<Station> stations = new ArrayList<>();
+        for (Section section : sections) {
+            stations.add(section.getUpStation());
+            stations.add(section.getDownStation());
+        }
+
+        return stations.stream()
+                .distinct()
+                .collect(Collectors.toList());
```

Comment:
sections.getLastStation() 메서드를 만들어야 하긴 하지만
아래처럼 하면 간단하게 표현 가능할 것 같아요 😄

```suggestion
       List<Station> stations = sections.stream()
                .map(it -> it.getUpStation())
                .collect(Collectors.toList());
        return stations.add(sections.getLastStation());
```